File: ms5-analytics/app/services/analytics_service.py
import logging
from collections import Counter, defaultdict
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.db_models import UserVideoEvent, UserVideoAnalytics
from app.models.schemas import (
    AnalyticsResponse,
    ImportantSection,
    RevisitedSegment,
    QueryEntry,
)
from app.services.highlight_service import generate_smart_highlights
from app.utils.time_utils import bucket_timestamp, merge_adjacent_buckets
from app.config import get_settings

logger = logging.getLogger(__name__)

settings = get_settings()

# Signal weights for importance scoring
SIGNAL_WEIGHTS = {
    "SEARCH": 5.0,
    "REPLAY": 3.0,
    "SEEK": 1.5,
    "PAUSE": 1.0,
    "PLAY": 0.0,
}


async def get_analytics(
    db: AsyncSession,
    user_id: str,
    video_id: str,
) -> AnalyticsResponse:
    """Retrieves or computes analytics for a user-video pair.

    Checks the analytics table first, otherwise computes from raw events.
    """
    # Check for existing computed analytics
    existing = await db.execute(
        select(UserVideoAnalytics).where(
            UserVideoAnalytics.user_id == user_id,
            UserVideoAnalytics.video_id == video_id,
        )
    )
    record = existing.scalar_one_or_none()
    if record and record.last_computed_at:
        return AnalyticsResponse(
            user_id=user_id,
            video_id=video_id,
            important_sections=[ImportantSection(**s) for s in (record.important_timestamps or [])],
            smart_highlights=record.smart_highlights or [],
            query_history=[QueryEntry(**q) for q in (record.query_history or [])],
            revisited_segments=[RevisitedSegment(**r) for r in (record.revisited_segments or [])],
            last_computed_at=record.last_computed_at,
        )

    # Compute fresh from raw events
    analytics = await compute_analytics(db, user_id, video_id)

    # Store the computed result
    await upsert_analytics(db, user_id, video_id, analytics)

    return analytics


async def compute_analytics(
    db: AsyncSession, user_id: str, video_id: str
) -> AnalyticsResponse:
    """Computes analytics by querying all events for a user-video pair."""
    result = await db.execute(
        select(UserVideoEvent)
        .where(
            UserVideoEvent.user_id == user_id,
            UserVideoEvent.video_id == video_id,
        )
        .order_by(UserVideoEvent.created_at)
    )
    events = result.scalars().all()

    if not events:
        return AnalyticsResponse(
            user_id=user_id,
            video_id=video_id,
            last_computed_at=datetime.now(timezone.utc),
        )

    bucket_size = settings.TIMESTAMP_BUCKET_SECONDS

    # 1. Bucket timestamps and compute scores
    bucket_events = defaultdict(list)
    for event in events:
        if event.timestamp_sec is not None:
            bucket = bucket_timestamp(event.timestamp_sec, bucket_size)
            bucket_events[bucket].append(event)

    scored_buckets = []
    for bucket, events_in_bucket in bucket_events.items():
        score = 0.0
        signals = set()
        for event in events_in_bucket:
            weight = SIGNAL_WEIGHTS.get(event.event_type, 0.0)
            score += weight
            if weight > 0:
                signals.add(event.event_type)
        scored_buckets.append({
            "bucket": bucket,
            "score": score,
            "signals": list(signals),
        })

    # 2. Sort by score and take top N buckets
    scored_buckets.sort(key=lambda b: b["score"], reverse=True)
    top_buckets = scored_buckets[: settings.IMPORTANT_SECTIONS_COUNT * 3]

    # 3. Merge adjacent buckets into sections
    sections = merge_adjacent_buckets(top_buckets, bucket_size, max_gap_buckets=2)

    # 4. Sort sections by score and create ImportantSection objects
    sections.sort(key=lambda s: s["total_score"], reverse=True)
    important_sections = []
    for rank, section in enumerate(
        sections[: settings.IMPORTANT_SECTIONS_COUNT], start=1
    ):
        label = _generate_section_label(section, bucket_events, bucket_size)
        important_sections.append(
            ImportantSection(
                rank=rank,
                start_sec=section["start_sec"],
                end_sec=section["end_sec"],
                label=label,
                score=round(section["total_score"], 1),
                signals=section["signals"],
            )
        )

    # 5. Find revisited segments
    revisited_segments = _find_revisited_segments(
        events, bucket_size, settings.MIN_REVISIT_COUNT
    )

    # 6. Extract query history
    query_history = _extract_query_history(events)

    # 7. Generate smart highlights
    smart_highlights = generate_smart_highlights(
        important_sections,
        bucket_events,
        bucket_size,
        settings.TOP_HIGHLIGHTS_COUNT,
    )

    now = datetime.now(timezone.utc)

    return AnalyticsResponse(
        user_id=user_id,
        video_id=video_id,
        important_sections=important_sections,
        smart_highlights=smart_highlights,
        query_history=query_history,
        revisited_segments=revisited_segments,
        last_computed_at=now,
    )


async def recompute_analytics(
    db: AsyncSession, user_id: str, video_id: str
) -> None:
    """Force recomputes analytics: recomputes from raw events and stores."""
    analytics = await compute_analytics(db, user_id, video_id)
    await upsert_analytics(db, user_id, video_id, analytics)


async def upsert_analytics(
    db: AsyncSession,
    user_id: str,
    video_id: str,
    analytics: AnalyticsResponse,
) -> None:
    """Upserts computed analytics into the user_video_analytics table.

    Uses a portable SELECT-then-INSERT/UPDATE pattern instead of
    PostgreSQL-specific ON CONFLICT, so this works on SQLite too.
    """
    now = datetime.now(timezone.utc)
    data = {
        "important_timestamps": [s.model_dump(mode="json") for s in analytics.important_sections],
        "smart_highlights": [h.model_dump(mode="json") for h in analytics.smart_highlights],
        "query_history": [q.model_dump(mode="json") for q in analytics.query_history],
        "revisited_segments": [r.model_dump(mode="json") for r in analytics.revisited_segments],
        "last_computed_at": now,
        "updated_at": now,
    }

    existing = await db.execute(
        select(UserVideoAnalytics).where(
            UserVideoAnalytics.user_id == user_id,
            UserVideoAnalytics.video_id == video_id,
        )
    )
    record = existing.scalar_one_or_none()

    if record:
        for key, value in data.items():
            setattr(record, key, value)
    else:
        record = UserVideoAnalytics(user_id=user_id, video_id=video_id, **data)
        db.add(record)

    await db.flush()
    logger.info(f"Upserted analytics for user={user_id} video={video_id}")


def _find_revisited_segments(
    events: list, bucket_size: int, min_count: int
) -> list[RevisitedSegment]:
    """Identifies segments replayed or seeked to multiple times."""
    replay_seeks = [
        e for e in events
        if e.event_type in ("REPLAY", "SEEK") and e.timestamp_sec is not None
    ]
    bucket_counts = Counter(
        bucket_timestamp(e.timestamp_sec, bucket_size) for e in replay_seeks
    )
    return [
        RevisitedSegment(
            start_sec=bucket,
            end_sec=bucket + bucket_size,
            replay_count=count,
        )
        for bucket, count in sorted(bucket_counts.items())
        if count >= min_count
    ]


def _extract_query_history(events: list) -> list[QueryEntry]:
    """Extracts all SEARCH events as query history."""
    return [
        QueryEntry(
            query_text=e.query_text or "",
            searched_at=e.created_at,
            result_timestamp_sec=e.timestamp_sec,
        )
        for e in events
        if e.event_type == "SEARCH" and e.query_text
    ]


def _generate_section_label(
    section: dict, bucket_events: dict, bucket_size: int
) -> str:
    """Generates a human-readable label for a section based on its signals."""
    start = section["start_sec"]
    end = section["end_sec"]
    search_queries = []

    for bucket_ts, events in bucket_events.items():
        if start <= bucket_ts < end:
            for event in events:
                if event.event_type == "SEARCH" and event.query_text:
                    search_queries.append(event.query_text)

    if search_queries:
        most_common = Counter(search_queries).most_common(1)[0][0]
        return most_common.capitalize()

    signals = section.get("signals", [])
    if "REPLAY" in signals:
        return "Frequently revisited section"
    if "SEEK" in signals:
        return "Frequently accessed section"
    if "PAUSE" in signals:
        return "Key moment"

    return "Important section"

File: ms5-analytics/src/services/analyticsService.js
/**
 * Analytics computation — port of Python analytics_service.py
 */
const { v4: uuidv4 } = require('uuid');
const { getPool } = require('../database');
const config = require('../config');
const { bucketTimestamp, mergeAdjacentBuckets } = require('../utils/timeUtils');
const { generateSmartHighlights } = require('./highlightService');

// Signal weights for importance scoring
const SIGNAL_WEIGHTS = {
  SEARCH: 5.0,
  REPLAY: 3.0,
  SEEK: 1.5,
  PAUSE: 1.0,
  PLAY: 0.0,
};

/**
 * Retrieves or computes analytics for a user-video pair.
 */
async function getAnalytics(userId, videoId) {
  const pool = getPool();

  // Check for existing computed analytics
  const existing = await pool.query(
    `SELECT * FROM user_video_analytics WHERE user_id = $1 AND video_id = $2`,
    [userId, videoId]
  );

  if (existing.rows.length && existing.rows[0].last_computed_at) {
    const record = existing.rows[0];
    return {
      user_id: userId,
      video_id: videoId,
      important_sections: record.important_timestamps || [],
      smart_highlights: record.smart_highlights || [],
      query_history: record.query_history || [],
      revisited_segments: record.revisited_segments || [],
      last_computed_at: record.last_computed_at,
    };
  }

  // Compute fresh from raw events
  const analytics = await computeAnalytics(userId, videoId);
  await upsertAnalytics(userId, videoId, analytics);
  return analytics;
}

/**
 * Computes analytics from raw events for a user-video pair.
 */
async function computeAnalytics(userId, videoId) {
  const pool = getPool();
  const result = await pool.query(
    `SELECT * FROM user_video_events
     WHERE user_id = $1 AND video_id = $2
     ORDER BY created_at`,
    [userId, videoId]
  );
  const events = result.rows;
  const now = new Date().toISOString();

  if (!events.length) {
    return {
      user_id: userId,
      video_id: videoId,
      important_sections: [],
      smart_highlights: [],
      query_history: [],
      revisited_segments: [],
      last_computed_at: now,
    };
  }

  const bucketSize = config.timestampBucketSeconds;

  // 1. Bucket timestamps and compute scores
  const bucketEvents = {}; // bucket_ts -> [events]
  for (const event of events) {
    if (event.timestamp_sec != null) {
      const bucket = bucketTimestamp(event.timestamp_sec, bucketSize);
      if (!bucketEvents[bucket]) bucketEvents[bucket] = [];
      bucketEvents[bucket].push(event);
    }
  }

  const scoredBuckets = [];
  for (const [bucket, eventsInBucket] of Object.entries(bucketEvents)) {
    let score = 0;
    const signals = new Set();
    for (const event of eventsInBucket) {
      const weight = SIGNAL_WEIGHTS[event.event_type] || 0;
      score += weight;
      if (weight > 0) signals.add(event.event_type);
    }
    scoredBuckets.push({
      bucket: Number(bucket),
      score,
      signals: [...signals],
    });
  }

  // 2. Sort by score, take top N*3 buckets
  scoredBuckets.sort((a, b) => b.score - a.score);
  const topBuckets = scoredBuckets.slice(0, config.importantSectionsCount * 3);

  // 3. Merge adjacent buckets into sections
  const sections = mergeAdjacentBuckets(topBuckets, bucketSize, 2);

  // 4. Sort sections by score and create ImportantSection objects
  sections.sort((a, b) => b.total_score - a.total_score);
  const importantSections = sections
    .slice(0, config.importantSectionsCount)
    .map((section, idx) => ({
      rank: idx + 1,
      start_sec: section.start_sec,
      end_sec: section.end_sec,
      label: generateSectionLabel(section, bucketEvents),
      score: Math.round(section.total_score * 10) / 10,
      signals: section.signals,
    }));

  // 5. Find revisited segments
  const revisitedSegments = findRevisitedSegments(events, bucketSize, config.minRevisitCount);

  // 6. Extract query history
  const queryHistory = extractQueryHistory(events);

  // 7. Generate smart highlights
  const smartHighlights = generateSmartHighlights(
    importantSections, bucketEvents, bucketSize, config.topHighlightsCount
  );

  return {
    user_id: userId,
    video_id: videoId,
    important_sections: importantSections,
    smart_highlights: smartHighlights,
    query_history: queryHistory,
    revisited_segments: revisitedSegments,
    last_computed_at: now,
  };
}

/**
 * Force recompute analytics from raw events and store.
 */
async function recomputeAnalytics(userId, videoId) {
  const analytics = await computeAnalytics(userId, videoId);
  await upsertAnalytics(userId, videoId, analytics);
}

/**
 * Upserts computed analytics into the user_video_analytics table.
 */
async function upsertAnalytics(userId, videoId, analytics) {
  const pool = getPool();
  const now = new Date().toISOString();
  const data = {
    important_timestamps: JSON.stringify(analytics.important_sections),
    smart_highlights: JSON.stringify(analytics.smart_highlights),
    query_history: JSON.stringify(analytics.query_history),
    revisited_segments: JSON.stringify(analytics.revisited_segments),
  };

  const existing = await pool.query(
    `SELECT id FROM user_video_analytics WHERE user_id = $1 AND video_id = $2`,
    [userId, videoId]
  );

  if (existing.rows.length) {
    await pool.query(
      `UPDATE user_video_analytics
       SET important_timestamps = $1, smart_highlights = $2, query_history = $3,
           revisited_segments = $4, last_computed_at = $5, updated_at = $5
       WHERE user_id = $6 AND video_id = $7`,
      [data.important_timestamps, data.smart_highlights, data.query_history,
       data.revisited_segments, now, userId, videoId]
    );
  } else {
    await pool.query(
      `INSERT INTO user_video_analytics
       (id, user_id, video_id, important_timestamps, smart_highlights,
        query_history, revisited_segments, last_computed_at, updated_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $8)`,
      [uuidv4(), userId, videoId, data.important_timestamps, data.smart_highlights,
       data.query_history, data.revisited_segments, now]
    );
  }

  console.log(`[analytics] Upserted analytics for user=${userId} video=${videoId}`);
}

function findRevisitedSegments(events, bucketSize, minCount) {
  const replaySeeks = events.filter(
    (e) => (e.event_type === 'REPLAY' || e.event_type === 'SEEK') && e.timestamp_sec != null
  );
  const bucketCounts = {};
  for (const e of replaySeeks) {
    const bucket = bucketTimestamp(e.timestamp_sec, bucketSize);
    bucketCounts[bucket] = (bucketCounts[bucket] || 0) + 1;
  }

  return Object.entries(bucketCounts)
    .filter(([, count]) => count >= minCount)
    .sort(([a], [b]) => Number(a) - Number(b))
    .map(([bucket, count]) => ({
      start_sec: Number(bucket),
      end_sec: Number(bucket) + bucketSize,
      replay_count: count,
    }));
}

function extractQueryHistory(events) {
  return events
    .filter((e) => e.event_type === 'SEARCH' && e.query_text)
    .map((e) => ({
      query_text: e.query_text,
      searched_at: e.created_at,
      result_timestamp_sec: e.timestamp_sec,
    }));
}

function generateSectionLabel(section, bucketEvents) {
  const searchQueries = [];
  for (const [bucketTs, events] of Object.entries(bucketEvents)) {
    const ts = Number(bucketTs);
    if (ts >= section.start_sec && ts < section.end_sec) {
      for (const event of events) {
        if (event.event_type === 'SEARCH' && event.query_text) {
          searchQueries.push(event.query_text);
        }
      }
    }
  }

  if (searchQueries.length) {
    const counts = {};
    searchQueries.forEach((q) => { counts[q] = (counts[q] || 0) + 1; });
    const mostCommon = Object.entries(counts).sort((a, b) => b[1] - a[1])[0][0];
    return mostCommon.charAt(0).toUpperCase() + mostCommon.slice(1);
  }

  const signals = section.signals || [];
  if (signals.includes('REPLAY')) return 'Frequently revisited section';
  if (signals.includes('SEEK')) return 'Frequently accessed section';
  if (signals.includes('PAUSE')) return 'Key moment';
  return 'Important section';
}

module.exports = { getAnalytics, recomputeAnalytics };
